File: api/src/margin_api/schemas/risk_diffing.py
"""Pydantic schemas for the risk factor diffing API."""

from __future__ import annotations

from datetime import date, datetime

from pydantic import BaseModel


class MaterialChangeResponse(BaseModel):
    change_type: str
    topic: str
    severity: int
    summary_50_words: str
    verbatim_new_text: str | None = None
    verbatim_old_text: str | None = None


class RiskFactorAnalysisResponse(BaseModel):
    ticker: str
    current_period: date
    prior_period: date
    overall_risk_delta_score: float
    model_confidence: float
    material_changes: list[MaterialChangeResponse]
    prompt_version: str
    analyzed_at: datetime

File: web/src/lib/api/risk_diffing.ts
/**
 * Risk factor diffing API client helpers.
 */
import { apiFetch } from "./client"

// ---------------------------------------------------------------------------
// Types matching Python Pydantic schemas for risk factor analysis
// ---------------------------------------------------------------------------

export interface MaterialChange {
  change_type: string
  topic: string
  severity: number
  summary_50_words: string
  verbatim_new_text: string
  verbatim_old_text: string
}

export interface RiskFactorAnalysis {
  ticker: string
  current_period: string
  prior_period: string
  overall_risk_delta_score: number
  model_confidence: number
  material_changes: MaterialChange[]
  prompt_version: string
  analyzed_at: string
}

// ---------------------------------------------------------------------------
// API functions
// ---------------------------------------------------------------------------

export async function getRiskFactorAnalysis(
  ticker: string,
): Promise<RiskFactorAnalysis | null> {
  try {
    return await apiFetch<RiskFactorAnalysis>(
      `/api/v1/analytics/risk_factors/${ticker.toUpperCase()}`,
    )
  } catch {
    return null
  }
}
